SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialNetDriverRPC.h

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "Interop/RPCs/RPCService.h"
+#include "SpatialView/EntityComponentId.h"
+#include "SpatialNetDriverRPC.generated.h"
+
+namespace SpatialGDK
+{
+class SpatialEventTracer;
+}
+
+/**
+ * Base RPC type
+ * Different from RPCPayload to have stricter definition of outgoing RPC data
+ * and to see if the templated version of the queue/sender/receiver is viable.
+ */
+struct FRPCPayload
+{
+	uint32 Offset;
+	uint32 Index;
+	TArray<uint8> PayloadData;
+
+	void ReadFromSchema(Schema_Object* RPCObject);
+	void WriteToSchema(Schema_Object* RPCObject) const;
+};
+
+/**
+ * Payload version for commands, adding a unique Id to try to prevent double-execution
+ * when double sending happens.
+ * NOTE : to have a uniform cross-server RPC API, the queue and sender would use a FRPCCrossServerPayload template parameter,
+ * but the sender would write a FRPCCommandPayload and the receiver would use a FRPCCommandPayload template parameter
+ */
+struct FRPCCommandPayload : FRPCPayload
+{
+	uint64 UUID;
+};
+
+/**
+ * Payload for cross-server RPCs.
+ * Depending on the component this payload is written to, Counterpart would indicate either the sender or the receiver.
+ * This is mainly intended to be used with the RoutingWorker implementation of cross-server RPCs,
+ * which is responsible for swapping the entity the RPC is written on and the counterpart.
+ */
+struct FRPCCrossServerPayload : FRPCPayload
+{
+	uint64 UUID;
+	Worker_EntityId Counterpart;
+};
+
+/**
+ * Additional data accompanying a received RPC.
+ */
+struct FRPCMetaData
+{
+	uint64 Timestamp;
+	FSpatialGDKSpanId SpanId;
+
+	void ComputeSpanId(FName Name, SpatialGDK::SpatialEventTracer& Tracer, SpatialGDK::EntityComponentId EntityComponent, uint64 RPCId);
+};
+
+/**
+ * Wrapper used to add Timestamp and event tracing information to a received RPC.
+ */
+template <typename T>
+struct TimestampAndETWrapper
+{
+	TimestampAndETWrapper(FName InRPCName, Worker_ComponentId InComponentId, SpatialGDK::SpatialEventTracer* InEventTracer = nullptr)
+		: RPCName(InRPCName)
+		, ComponentId(InComponentId)
+		, EventTracer(InEventTracer)
+	{
+	}
+
+	using AdditionalData = FRPCMetaData;
```

Comment:
Do you use this "using"?

Reply:
This is used in the template definitions of the RPC queues.